### PR TITLE
persist: reduce S3 read timeout

### LINKED ISSUE #17227: Queries sometimes hang for an additional minute
### What version of Materialize are you using?

main

### How did you install Materialize?

Materialize Cloud

### What is the issue?

Interactive queries that usually complete fairly quickly sometimes instead take an additional minute to complete. The cause of this seem to be timeouts during fetching blobs from S3 in persist.

Specifically, I noticed this while running TPC-H benchmarks. It is possible that this issue occurs more when large amounts of data are involved.

### Reproduction

I could only reproduce this in cloud (staging), not locally. It might reproduce locally as well, if persist was configured to use S3 for the blob store.

Run the following queries to set up the TPC-H sources and a test cluster:
```
CREATE SOURCE lgtpch FROM LOAD GENERATOR TPCH (SCALE FACTOR 5, TICK INTERVAL '1s') FOR ALL TABLES WITH (SIZE = 'xsmall');

CREATE CLUSTER test REPLICAS (r1 (SIZE 'large'));
SET cluster = test;
```

Now wait a bit until the TPC-H sources have caught up.

Define a view for TPC-H query 2. Others will probably work as well.

```
CREATE VIEW q02 AS SELECT
    s_acctbal,
    s_name,
    n_name,
    p_partkey,
    p_mfgr,
    s_address,
    s_phone,
    s_comment
FROM
    part, supplier, partsupp, nation, region
WHERE
    p_partkey = ps_partkey
    AND s_suppkey = ps_suppkey
    AND p_size = CAST (15 AS smallint)
    AND p_type LIKE '%BRASS'
    AND s_nationkey = n_nationkey
    AND n_regionkey = r_regionkey
    AND r_name = 'EUROPE'
    AND ps_supplycost
        = (
                SELECT
                    min(ps_supplycost)
                FROM
                    partsupp, supplier, nation, region
                WHERE
                    p_partkey = ps_partkey
                    AND s_suppkey = ps_suppkey
                    AND s_nationkey = n_nationkey
                    AND n_regionkey = r_regionkey
                    AND r_name = 'EUROPE'
            )
ORDER BY
    s_acctbal DESC, n_name, s_name, p_partkey;
```

Repeatedly run the query until the described issue occurs:
```
materialize=> \timing
Timing is on.
materialize=> \o /dev/null
materialize=> SELECT * FROM q02;
Time: 11262.504 ms (00:11.263)
materialize=> SELECT * FROM q02;
Time: 11282.394 ms (00:11.282)
materialize=> SELECT * FROM q02;
Time: 65270.738 ms (01:05.271)
```

### Initial Investigation

I have traced this down to blob fetching from S3. In particular, if you add logging around this part of the code:
https://github.com/MaterializeInc/materialize/blob/845f8ed2aca78f86638053a106cc56d5da98d411/src/persist-client/src/fetch.rs#L236-L240

... you see that it blocks for about one minute:

```
│ clusterd {"timestamp":"2023-01-19T14:42:09.610269Z","level":"INFO","fields":{"message":"[...] fetching..."},"target":"mz_persist_client::fetch"}          │
│ clusterd {"timestamp":"2023-01-19T14:43:10.804087Z","level":"INFO","fields":{"message":"[...] fetching... done"},"target":"mz_persist_client::fetch"}     │
```

That one minute suspiciously matches [the read timeout we configure with the S3 SDK](https://github.com/MaterializeInc/materialize/blob/845f8ed2aca78f86638053a106cc56d5da98d411/src/persist/src/s3.rs#L105). So a likely explanation for the above behavior is that reading the blob from S3 fails for some time, times our after 60 seconds, and immediately succeeds on the second try.

Relevant discussion [in Slack](https://materializeinc.slack.com/archives/C01CFKM1QRF/p1674140180968459).

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by teskje**
It appears to be a common occurrence that reads from S3 time out the on the first try but immediately succeed on the second try. The previous S3 read timeout was 60 seconds, which lead to cases where reads from persist sometimes took a minute longer than usual, which in turn would become very noticable in one-off queries reading from a storage source.

This PR attempts to reduce the impact to query times by aggressively reducing the S3 read timeout. The new timeout is 4s, which would still introduce a noticable delay, especially to otherwise fast queries, but much less than the previous timeout.

4s seems like a safe timeout, since our persist clients are expected to be located physically close to the S3 endpoints. We might even be able to go lower than that after some testing.

### Motivation

Which of the following best describes the motivation behind this PR?

  * This PR fixes a recognized bug.

Fixes #17227.

### Tips for reviewer

I understand that the persist teams plans a rethink of how our S3 timeouts work. This PR is not meant to be that. It is just a short-term fix for #17227.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a
  companion cloud PR to account for those changes that is tagged with
  the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - N/A

### CODE REVIEW

**Comment by teskje** [comment]
I tested the effects of this change in my staging account, with the repro described in #17227. After 270 queries, the maximum query time observed was ~7.5s. So I would say the fix is effective!

File: src/persist/src/s3.rs
Comment (pH14) [inline src/persist/src/s3.rs:103]: I wouldn't mind slightly higher on this one, since we don't know our typical connect timings right now. 2s is probably fine, but I'd feel a little more comfortable bumping to maybe 3-4s until we have more metric coverage

**Review by pH14 — APPROVED** [review]
Beat me to it! Thank you @teskje 🙇

File: src/persist/src/s3.rs
Comment (teskje) [reply src/persist/src/s3.rs:103]: This might be wrong, but the way I read [the docs](https://docs.rs/aws-config/0.53.0/aws_config/timeout/struct.TimeoutConfig.html#method.read_timeout) the `read_timeout` includes the time required to connect, so the read timeout should always be greater than the  connect timeout.

So if we want to bump the connect timeout to 4s, and still allow for 2s to read something from the connection, we should set the read timeout to 6s. Does that make sense?

Btw, these values are total guesswork. Happy to increase them until we all feel comfortable!

**Comment by danhhz** [comment]
Oooo thanks for the fix, Jan! When this came up, I started a discussion within the team about how we should prioritize these sorts of issues, and then... we got distracted from the issue itself 🙈, sorry!

Don't wait for my approval, I'm happy when Paul is

File: src/persist/src/s3.rs
Comment (pH14) [reply src/persist/src/s3.rs:103]: maybe 3 for connect and 6 for read then? that'd make me a little more comfortable when first rolling this out. I have a PR incoming that'll add some metric coverage + make the timeouts dynamic (once LD is hooked in), so we can lower them further more safely afterwards

**Comment by pH14** [comment]
Oops, I think we'll merge conflict between this branch and https://github.com/MaterializeInc/materialize/pull/17246 where it makes sense to just merge the latter. I'll get it in after the nightlies finish on it!

**Comment by teskje** [comment]
Sure, feel free to merge yours and close this one! I've already signed off for the weekend and won't be doing anything until Monday.

**Comment by antiguru** [comment]
Could we turn the various timeouts into configuration flags? This would allow us to address any surprising effects of this change other than by rolling out a new release.

**Comment by teskje** [comment]
Looks like that is what #17246 is doing!

**Comment by pH14** [comment]
https://github.com/MaterializeInc/materialize/pull/17246 has been merged and brings the connect/read timeouts down to 7 and 10s. We can play with those numbers more aggressively once we had LD hooked up. Thanks for bringing this one to our attentino @teskje!

**Pull request closed without merging**
